src/Functions/FunctionsExternalDictionaries.h

```diff
@@ -594,31 +615,122 @@ class FunctionDictGetNoType final : public IFunction
+        ColumnPtr & result_column,
+        ColumnPtr defaults_column,
+        ColumnPtr mask_column,
+        IColumn::Filter & inverted_mask,
+        const DataTypePtr & result_type) const
+    {
+        auto mut_lhs = IColumn::mutate(std::move(result_column));
+        mut_lhs->expand(inverted_mask, true);
+
+        auto if_func = FunctionFactory::instance().get("if", helper.getContext());
+        ColumnsWithTypeAndName if_args =
+        {
+            {mask_column, std::make_shared<DataTypeUInt8>(), {}},
+            {std::move(mut_lhs), result_type, {}},
+            {defaults_column, result_type, {}},
+        };
+
+        auto rows = inverted_mask.size();
+        result_column = if_func->build(if_args)->execute(if_args, result_type, rows);
+    }
+
+#ifdef ABORT_ON_LOGICAL_ERROR
+    static void validateShortCircuitResult(const ColumnPtr & column, const IColumn::Filter & filter)
+    {
+        size_t expected_size = countBytesInFilter(filter);
+        size_t col_size = column->size();
+        if (col_size != expected_size)
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Invalid size of getColumnsOrDefaultShortCircuit result. Column has {} rows, but filter contains {} bytes.",
+                col_size, expected_size);
+    }
+#endif
+
     ColumnPtr executeDictionaryRequest(
         std::shared_ptr<const IDictionary> & dictionary,
         const Strings & attribute_names,
         const Columns & key_columns,
         const DataTypes & key_types,
-        const DataTypePtr & result_type,
+        const DataTypePtr & attribute_type,
         const Columns & default_cols,
-        size_t collect_values_limit) const
+        size_t collect_values_limit,
+        const ColumnWithTypeAndName & last_argument,
+        const DataTypePtr & result_type) const
     {
         ColumnPtr result;
 
         if (attribute_names.size() > 1)
         {
-            const auto & result_tuple_type = assert_cast<const DataTypeTuple &>(*result_type);
+            const auto & attribute_tuple_type = assert_cast<const DataTypeTuple &>(*attribute_type);
 
             Columns result_columns;
             if constexpr (dictionary_get_function_type == DictionaryGetFunctionType::getAll)
             {
                 result_columns = dictionary->getColumnsAllValues(
-                    attribute_names, result_tuple_type.getElements(), key_columns, key_types, default_cols, collect_values_limit);
+                    attribute_names, attribute_tuple_type.getElements(), key_columns, key_types, default_cols, collect_values_limit);
+            }
+            else if (dictionary_get_function_type == DictionaryGetFunctionType::getOrDefault && default_cols.empty())
+            {
+                IColumn::Filter default_mask;
+                result_columns = dictionary->getColumnsOrDefaultShortCircuit(
+                    attribute_names, attribute_tuple_type.getElements(), key_columns, key_types, default_mask);
```

Comment:
As I understand, `default_mask` will contain `0` for rows with default values. Is it intentional? For me it sounds more natural to make in `1` for rows with default values.